src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+};
+
+typedef struct WalFilter
+{
+    ReadStep currentStep;
+    const unsigned char *page;
+
+    PgPageSize heapPageSize;
+    PgPageSize walPageSize;
+    uint32_t segSize;
+
+    bool is_begin;
+
+    size_t page_offset;
+    size_t input_offset;
+    size_t total_offset;
+
+    XLogPageHeaderData *current_header;
+
+    XLogRecord *record;
+    size_t rec_buf_size;
+    // Offset to the body of the record on the current page
+    size_t header_offset;
+    // How many we read of this record
+    size_t got_len;
+    // Total size of record on current page
+    size_t tot_len;
+
+    List *headers;
+
+    WalInterface *walInterface;
+
+    const ArchiveGetFile *archiveInfo;
+
+    // Records count for debug
+    uint32_t i;
+
+    bool done;
+    bool same_input;
+    bool is_switch_wal;
+} WalFilterState;
+
+/***********************************************************************************************************************************
+Render as string for logging
+***********************************************************************************************************************************/
+static void
+walFilterToLog(const WalFilterState *const this, StringStatic *const debugLog)
+{
+    strStcFmt(
+        debugLog,
+        "{record_num: %u, step: %u is_begin: %s, page_offset: %zu, input_offset: %zu, rec_buf_size: %zu,"
+        " header_offset: %zu, got_len: %zu, tot_len: %zu}",
+        this->i,
+        this->currentStep,
+        this->is_begin ? "true" : "false",
+        this->page_offset,
+        this->input_offset,
+        this->rec_buf_size,
+        this->header_offset,
+        this->got_len,
+        this->tot_len
+        );
+}
+
+#define FUNCTION_LOG_WAL_FILTER_TYPE                                                                                               \
+    ZstDecompress *
+#define FUNCTION_LOG_WAL_FILTER_FORMAT(value, buffer, bufferSize)                                                                  \
+    FUNCTION_LOG_OBJECT_FORMAT(value, walFilterToLog, buffer, bufferSize)
+
+static inline
+void
+checkOutputSize(Buffer *const output, const size_t size)
+{
+    if (bufRemains(output) <= size)
+    {
+        bufResize(output, bufUsed(output) + size);
+    }
+}
+
+// Reading the next page from the input buffer. If the input buffer is exhausted, remember the current step and returns true.
+// In this case, we should exit the process function to get a new input buffer.
+static inline
+bool
+readPage(WalFilterState *const this, const Buffer *const input, const ReadStep step)
+{
+    if (this->input_offset >= bufUsed(input))
+    {
+        ASSERT(step != 0);
+        this->currentStep = step;
+        this->input_offset = 0;
+        this->same_input = false;
+        return true;
+    }
+    this->page = bufPtrConst(input) + this->input_offset;
+    this->input_offset += this->walPageSize;
+    this->page_offset = 0;
+    this->currentStep = 0;
```

Comment:
What does 0 mean for step value? It is not in the enumeration! How did the compiler allow this?!